### ISSUE TITLE
Consistent Variable Names in Snippets

### ISSUE THREAD

**Issue by jonashae** [muthesius/pepe#4]
Sollte eine sinnvolle Convention geben; für Snippet Entwickler

**Comment by jonashae** [comment]
Punkte die bis jetzt beachtet werden müssen, bzw. was es bis jetzt gibt:

Um auf bestimmte Positionen im Arduino Code zu kommen:

```
VAVA.PRESETUP
*alles was vor dem void Setup kommt.*

VAVA.SETUP
*alles was ins void Setup kommt*

VAVA.VARIABLE
*Variablen die ins void Loop kommen*

VAVA.SCRIPT 
*Dinge die im void Loop ausgeführt werden können*

VAVA.FUNKTION
*Platz nach void Loop - kann für Funktionen verwendet werden*
```

Die Trennung zwischen Script und Variablen macht auf den ersten Blick nicht wirklich viel Sinn und ist wohl der der Neuausschreibung mit der NodeID geschuldet - liest sich aber meiner Meinung nach ganz angenehm:

```
VAVA.SETUP
pinMode({{Pin}},INPUT_PULLUP);
pinMode({{PinLamp}}, OUTPUT);

VAVA.VARIABLE 
int sensorVal{{Pin}} = digitalRead({{Pin}});

VAVA.SCRIPT
  if (sensorVal{{Pin}} == HIGH) {
    digitalWrite({{PinLamp}}, LOW);
  } 
  else {
    digitalWrite({{PinLamp}}, HIGH);
  }
VAVA
```

Sollen wir das so weiterhin übernehmen? hätte dann alles in den finalen Namen geändert (pepe.script etc.) und würde vielleicht hier noch die OFF States einfügen also evt. einmal

```
VAVA.VARIABLE_On
int sensorVal{{Pin}} = digitalRead({{Pin}});

VAVA.VARIBALE_OFF
int sensorVal{{Pin}} = HIGH;
digitalWrite({{PinLamp}}, LOW);
```

So das Variablen und Befehle ausgeführt werden können - bin mir nur nicht sicher ob dann die Bezeichung Varible_off Sinn macht oder eher was in die Richtung: Case_OFF, Reset_State etc. ?

Bezeichung für eigentliche Variablen ist noch interessant - eigentlich brauchen wir das nicht festzulegen weil die eh schnell in anderen Namen geändert werden, damit man aber die Scripte von anderen schneller lesen ist macht das wohl Sinn die besonders zu Kennzeichnen - ich glaube nur das {{Vari}} vielleicht nicht so gut ist?

**Comment by jnslxndr** [comment]
OK. Das ist schon ein WIKI-reifer Text! Sehr gut dokumentiert!
Letzter Part ist für mich nicht ganz nachvollziehbar. Unter Umständen reichen die Teile `BASE` (a.k.a. `PRESETUP`, aber klarer), `SETUP` und `SCRIPT` aus. Wenn jemand Funktionen deklarieren will, dann kann er das in `BASE` ebenso tun; ist vom C-Code her nicht relevant (bzw. sogar korrekter).
Was zu beachten wäre, ist der Fall der Pin-Substituierung! wird diese im gesamten Schnipsel gemacht oder nur in `BASE`?

Hier muss ich auch noch mal wegen des Regex-Mappings fragen: Wie sinnvoll wäre eine Syntax à la `{{PIN_<slice index im pin spread>}}` ?

Übrigens ein Tip: Wenn man Variablen im globalen Scope definiert und sich diese nicht ändern, können sie ruhig mit `const int Pin_{{PIN}} …` erstellt werden. Kleine Optimierung…

**Comment by jonashae** [comment]
Die Pin-Substituierung, wenn ich deine Frage richtig verstehe, nur im Base gemacht. 

```
int Digital_04 = 4;
int Digital_12 = 12;

void setup(){
pinMode(Digital_12,INPUT_PULLUP);
pinMode(Digital_04, OUTPUT);
}
void loop() {
int sensorVal_NodeID1170Digital_12 = digitalRead(Digital_12);
if (sensorVal_NodeID1170Digital_12 == HIGH) {
    digitalWrite(Digital_04, LOW);
  } 
  else {
    digitalWrite(Digital_04, HIGH);
  }
}
```

Im Base werden die möglichen Enums auf die Pins definiert, im späteren soll eigentlich nur noch mit den Enum Variablen und nicht mit den Pin Bezeichnungen gearbeitet werden.

Eine Syntax mit Slice Indexen hatte ich schon drinnen, da ich allerdings die Enums dynamisch erstelle zerhackt es mit die Indexe - daher hab ich mich entschieden die Enums wirklich als Variablen zu verwenden:

![image](https://cloud.githubusercontent.com/assets/4255191/6440368/96d3bca0-c0dc-11e4-9f5b-0d0ec0e1c543.png)

Hier kurz als Beispiel - wenn ich den Pin bzw. die Enum Pin2 als Input verwende wird diese aus der Liste für mögliche Outputs gelöscht. Wenn man Mist baut - Inputs kopiert etc passiert das hier:

![image](https://cloud.githubusercontent.com/assets/4255191/6440438/20eb095c-c0dd-11e4-8587-b91d6da7ab5f.png)

Im Moment zickt das System rum (dynamische Plugins sind nicht meine Stärke ;-))  aber wenn es fehlerfrei läuft kann der Nutzer später keine Input mit Outputs gleichschalten, und wenn doch (Enum kopieren) wird die Node fehlerhaft angezeigt, was das System leichter verständlich macht.

**Comment by jnslxndr** [comment]
ja, das klärt die frage. ist aus meiner sicht auch gut so, im `BASE`.
was mir dabei auffällt: wäre es nicht auch denkbar eine feste PinListe zu haben, sodass man das nicht mehr mit dem Regexp rausgfischen muss. Die Enum-Nodes machen da aber nach, wie vor Sinn!

**Comment by jonashae** [comment]
? ähm welches regexp meinst du? das für die schnippsel ist zwingend (naja substitute täte es auch) nötig das ich ja einen Quelltext schreiben will, indem ich in Anschluss die Pins festlegen kann - sonst hätte ich ja das Problem das meine LED Lampe nur mit einem bestimmten Pin läuft - oder stehe ich grade auf dem Schlauch

**Comment by jnslxndr** [comment]
stimmt, ersetzen muss man es an einem punkt. ich dachte an eine festgelegte pinliste (zb. die namen der enum einträge), an die man sich als schnipsel-schreiber halten muss und somit keine eigenen namen festlegen kann/muss. vl. hilft das aber auch gar nicht, das kann ich nicht wirklich beurteilen :)
